packages-explorer/src/state.js

```diff
@@ -0,0 +1,354 @@
+ */
+const INITIAL_STATE = {
+	attr: null,
+	channel: null,
+	page: 1,
+	query: "",
+	unfree: false,
+};
+
+/**
+ * App's state and callbacks.
+ *
+ * This is built in a way that forces components to declare what they want to use
+ * from the state. Once declared, it is made available as props.
+ *
+ * See this as a poor man's redux store, but with tighter control.
+ *
+ * Some of this app's state is synchronized in the URL.
+ *
+ * Everything else is internal (large data mainly).
+ */
+class State extends Component {
+	constructor() {
+		super();
+
+		this.state = Object.assign({
+			loading: 0,
+			channels: [],
+			channel_data: null,
+			filtered_packages: [],
+			current_results: [],
+		}, INITIAL_STATE);
+
+		// Binding functions to `this` for use as callbacks.
+		// `handle_popstate` is used to link history with state.
+		["handle_popstate"]
+			.concat(CALLBACKS)
+			.forEach((fn) => { this[fn] = this[fn].bind(this); });
+
+		this.refilter = debounce(this.refilter, DEBOUNCE);
+	}
+
+	handle_popstate(e) {
+		let state = {};
+		if (e.state) {
+			state = e["state"];
+		}
+		else {
+			state = queryString.parse(location.search);
+		}
+		const new_state = fromPairs(SYNCHRONIZED.map((name) => [name, INITIAL_STATE[name]]));
+
+		this.setState(Object.assign(new_state, state), {push: false});
+	}
+
+	componentWillMount() {
+		const params = queryString.parse(location.search);
+		const {history} = window;
+		const state = Object.assign({}, params, history.state);
+		this.setState(state);
+
+		this.fetch_channels()
+			.then(() => {
+				this.fetch_channel();
+			})
+		;
+		window.addEventListener("popstate", this.handle_popstate);
+	}
+
+	setState(new_state, {push = true, force = false} = {}, ...args) {
+		const {history} = window;
+		const params = this.url_params_for_state(new_state);
+
+		// Applies the state change.
+		super.setState(new_state, ...args);
+
+		// History modification won't fire on "identity" change.
+		if (!force && isEqual(params, queryString.parse(window.location.search))) {
+			return;
+		}
+
+		let method = push ? "pushState" : "replaceState";
+
+		// Assumes that from no querystring to querystring is replace.
+		// This saves one "weird" state replacement.
+		if (Object.keys(params).length > 0 && window.location.search.length === 0) {
+			method = "replaceState";
+		}
+
+		if (queryString.stringify(params).length > 0) {
+			history[method](params, "", `?${queryString.stringify(params)}`);
+		}
+		else {
+			history[method](params, "", window.location.pathname);
+		}
+	}
+
+	componentDidUpdate(prev_props, prev_state) {
+		if (prev_state["channel"] !== this.state["channel"]) {
+			this.fetch_channel();
+
+			return;
+		}
+
+		const {state} = this;
+		if (
+			state["query"] !== prev_state["query"] ||
+			state["unfree"] !== prev_state["unfree"]
+		) {
+			this.refilter();
+		}
+	}
+
+	fetch_channels() {
+		this.setState({loading: this.state.loading + 1});
+
+		return fetch("/nixpkgs/packages-channels.json", {mode: "cors"})
```

Comment:
this file can change over time right? seems like you'd want hash here to cache bust this file when the channel has been updated, otherwise I think the browser will use the cached version

same applies to the below json file